src/common/src/hash/key_v2.rs

```diff
@@ -0,0 +1,389 @@
+
+                Either::Left(
+                    estimated_key_sizes
+                        .into_iter()
+                        .map(S::Buffer::with_capacity),
+                )
+            } else {
+                Either::Right((0..data_chunk.capacity()).map(|_| S::Buffer::with_capacity(0)))
+            };
+
+            for_both!(buffers, buffers =>
+                hash_codes
+                    .into_iter()
+                    .zip_eq_fast(buffers)
+                    .map(|(hash_code, buffer)| Serializer::new(buffer, hash_code))
+                    .collect_vec()
+            )
+        };
+
+        for &i in column_indices {
+            let array = data_chunk.column_at(i).array_ref();
+
+            // Dispatch types once to accelerate the inner call.
+            dispatch_all_variants!(array, ArrayImpl, array, {
+                for (scalar, serializer) in array.iter().zip_eq_fast(&mut serializers) {
+                    serializer.serialize(scalar);
+                }
+            });
+        }
+
+        let hash_keys = serializers.into_iter().map(|s| s.finish()).collect();
+        Ok(hash_keys)
+    }
+
+    fn deserialize(&self, data_types: &[DataType]) -> ArrayResult<OwnedRow> {
+        let mut deserializer = Deserializer::<S, N>::new(&self.key, &self.null_bitmap);
+        let mut row = Vec::with_capacity(data_types.len());
+
+        for data_type in data_types {
+            let datum = deserializer.deserialize_impl(data_type);
+            row.push(datum);
+        }
+
+        Ok(OwnedRow::new(row))
+    }
+
+    fn deserialize_to_builders(
+        &self,
+        array_builders: &mut [ArrayBuilderImpl],
+        data_types: &[DataType],
+    ) -> ArrayResult<()> {
+        let mut deserializer = Deserializer::<S, N>::new(&self.key, &self.null_bitmap);
+
+        for (data_type, array_builder) in data_types.iter().zip_eq_fast(array_builders.iter_mut()) {
+            // Dispatch types once to accelerate the inner call.
+            dispatch_all_variants!(array_builder, ArrayBuilderImpl, array_builder, {
+                let datum = deserializer.deserialize(data_type);
+                array_builder.append_owned(datum);
+            });
+        }
+
+        Ok(())
+    }
+
+    fn null_bitmap(&self) -> &Self::Bitmap {
+        &self.null_bitmap
+    }
+}
+
+pub type FixedSizeKey<const N: usize, B> = HashKeyImpl<StackStorage<N>, B>;
+pub type Key8<B = HeapNullBitmap> = FixedSizeKey<1, B>;
+pub type Key16<B = HeapNullBitmap> = FixedSizeKey<2, B>;
+pub type Key32<B = HeapNullBitmap> = FixedSizeKey<4, B>;
+pub type Key64<B = HeapNullBitmap> = FixedSizeKey<8, B>;
+pub type Key128<B = HeapNullBitmap> = FixedSizeKey<16, B>;
+pub type Key256<B = HeapNullBitmap> = FixedSizeKey<32, B>;
+pub type KeySerialized<B = HeapNullBitmap> = SerializedKey<B>;
+
+pub type SerializedKey<B = HeapNullBitmap> = HashKeyImpl<HeapStorage, B>;
```

Comment:
Wonder if it will be confusing to have `KeySerialized` and `SerializedKey` 😆 . But I don't have a better idea now.

Reply:
These're copied from the original implementation. This PR aims to keep all of the interfaces untouched so the changes are all in the `common` crate. Let's refactor the code structure along with this in next PRs.